File: composeApp/src/androidMain/kotlin/io/github/a7asoft/ui/ReaderViewModel.kt
package io.github.a7asoft.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import io.github.a7asoft.nfc.NfcAvailability
import io.github.a7asoft.nfc.NfcAvailabilityStatus
import io.github.a7asoft.nfcemv.reader.IoReason
import io.github.a7asoft.nfcemv.reader.ReaderError
import io.github.a7asoft.nfcemv.reader.ReaderState
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import java.util.concurrent.atomic.AtomicReference

/**
 * Owns the [ReaderUiState] for the reader screen.
 *
 * Plain [ViewModel] (not [androidx.lifecycle.AndroidViewModel]) so it
 * can be unit-tested without Robolectric. Platform-specific dependencies
 * ([NfcAvailability], the [ReaderHandle] per read) are injected via the
 * [provideFactory] builder and the public API.
 *
 * Concurrency contract: only one read flow runs at a time. A new
 * [onReadRequested] cancels any in-flight collection before starting
 * the new one (defensive against rapid taps).
 */
public class ReaderViewModel internal constructor(
    private val nfcAvailability: NfcAvailability,
) : ViewModel() {

    private val _state: MutableStateFlow<ReaderUiState> = MutableStateFlow(initialState())

    /** Current UI state. UI cannot mutate; only [onReadRequested] / [reset] / [refreshNfcAvailability] do. */
    public val state: StateFlow<ReaderUiState> = _state.asStateFlow()

    private val activeRead: AtomicReference<Job?> = AtomicReference(null)

    /**
     * Re-check NFC availability — call from `MainActivity.onResume`.
     *
     * Only refreshes when state is one of the resettable gates
     * ([ReaderUiState.NfcUnavailable], [ReaderUiState.NfcDisabled],
     * [ReaderUiState.Idle]). In-flight reads (TagDetected … ReadingRecords)
     * and terminal states (Done, Failed) are preserved so the user does
     * not lose progress or results when returning from system settings.
     */
    @Suppress("CyclomaticComplexMethod")
    // why: 3-branch sealed dispatch over the resettable subset of
    // [ReaderUiState]; detekt counts the function itself as +1, tipping
    // the threshold. The suppression matches that algorithm.
    public fun refreshNfcAvailability() {
        val current = _state.value
        val refreshable = current is ReaderUiState.NfcUnavailable ||
            current is ReaderUiState.NfcDisabled ||
            current is ReaderUiState.Idle
        if (refreshable) _state.value = initialState()
    }

    /**
     * Begin a read using [handle]. Cancels any in-flight read first.
     *
     * Uses an atomic Idle → TagDetected transition to guard against
     * concurrent invocations from NFC reader-mode binder-thread callbacks
     * (a TOCTOU race that a plain `if (state is Idle)` check cannot
     * defend against).
     */
    public fun onReadRequested(handle: ReaderHandle) {
        // why: atomic gate. NFC reader-mode callbacks fire on a binder
        // thread, not main; two concurrent taps could both pass a naive
        // `is Idle` check. compareAndSet is the smallest defensible fix.
        // The flow's first emission may overwrite TagDetected with the
        // same TagDetected (or the next state) — UI effect is identical.
        if (!_state.compareAndSet(ReaderUiState.Idle, ReaderUiState.TagDetected)) return
        activeRead.getAndSet(null)?.cancel()
        val job = handle.read()
            .onEach { state -> _state.value = mapReaderState(state) }
            // why: defensive. Any unexpected upstream throw should land as
            // Failed, not propagate to viewModelScope's exception handler.
            .catch { _state.value = ReaderUiState.Failed(ReaderError.IoFailure(IoReason.Generic)) }
            .launchIn(viewModelScope)
        activeRead.set(job)
    }

    /** Reset to the initial state after a Done/Failed terminal. */
    public fun reset() {
        activeRead.getAndSet(null)?.cancel()
        _state.value = initialState()
    }

    override fun onCleared() {
        activeRead.getAndSet(null)?.cancel()
        super.onCleared()
    }

    @Suppress("CyclomaticComplexMethod")
    // why: 3-branch sealed dispatch over [NfcAvailabilityStatus]. detekt's
    // branch-counting algorithm re-flags this at threshold 4; the
    // suppression matches that algorithm.
    private fun initialState(): ReaderUiState = when (nfcAvailability.current()) {
        NfcAvailabilityStatus.Unavailable -> ReaderUiState.NfcUnavailable
        NfcAvailabilityStatus.Disabled -> ReaderUiState.NfcDisabled
        NfcAvailabilityStatus.Available -> ReaderUiState.Idle
    }

    @Suppress("CyclomaticComplexMethod")
    // why: exhaustive `when` over the sealed `ReaderState` catalogue is the
    // project idiom (CLAUDE.md §3.2). Each branch is a single mapping, not
    // an arm of real cyclomatic complexity.
    private fun mapReaderState(state: ReaderState): ReaderUiState = when (state) {
        is ReaderState.TagDetected -> ReaderUiState.TagDetected
        is ReaderState.SelectingPpse -> ReaderUiState.SelectingPpse
        is ReaderState.SelectingAid -> ReaderUiState.SelectingApplication(state.aid)
        is ReaderState.ReadingRecords -> ReaderUiState.ReadingRecords
        is ReaderState.Done -> ReaderUiState.Done(state.card)
        is ReaderState.Failed -> ReaderUiState.Failed(state.error)
    }

    public companion object {
        /**
         * Build the production-wired [ViewModelProvider.Factory].
         *
         * @param availability NFC capability checker (typically
         *   [io.github.a7asoft.nfc.SystemNfcAvailability]).
         */
        public fun provideFactory(
            availability: NfcAvailability,
        ): ViewModelProvider.Factory = viewModelFactory {
            initializer { ReaderViewModel(availability) }
        }
    }
}

File: iosApp/iosApp/ui/ReaderViewModel.swift
import Foundation
import EmvReader
import Shared

/// Owns the ``ReaderUiState`` for the reader screen.
///
/// Decouples the SwiftUI view tree from the `EmvReader` AsyncStream
/// lifecycle. Mirrors the Android `ReaderViewModel` (composeApp #55):
/// platform dependencies are injected via initialiser closures so the
/// type can be unit-tested without a simulator.
///
/// Concurrency: the read task runs on the main actor — every state
/// transition happens on the UI thread, matching SwiftUI's expectations
/// for `@Published` mutations.
@MainActor
public final class ReaderViewModel: ObservableObject {

    /// Current UI state. Drives ``ReaderScreen`` rendering. UI cannot
    /// mutate; only ``startScan()``, ``reset()``, and
    /// ``refreshAvailability()`` do.
    @Published public private(set) var uiState: ReaderUiState

    private let availability: NfcAvailability
    private let readerFactory: () -> AsyncStream<ReaderState>
    private var activeTask: Task<Void, Never>?

    /// Build a view-model with explicit dependencies.
    ///
    /// - Parameters:
    ///   - availability: NFC capability checker. Production:
    ///     ``SystemNfcAvailability``.
    ///   - readerFactory: Factory producing a fresh
    ///     `AsyncStream<ReaderState>` per scan. Production:
    ///     `{ EmvReader().read() }`.
    public init(
        availability: NfcAvailability,
        readerFactory: @escaping () -> AsyncStream<ReaderState>
    ) {
        self.availability = availability
        self.readerFactory = readerFactory
        self.uiState = Self.initialState(availability: availability)
    }

    /// Re-check NFC availability — call from `.onAppear`.
    ///
    /// Only refreshes when state is one of the resettable gates
    /// (``ReaderUiState/nfcUnavailable`` or ``ReaderUiState/idle``).
    /// In-flight reads and terminal states (``ReaderUiState/done(card:)``,
    /// ``ReaderUiState/failed(error:friendlyMessage:diagnostic:)``) are
    /// preserved so the user does not lose progress or results.
    public func refreshAvailability() {
        if isResettable(uiState) {
            uiState = Self.initialState(availability: availability)
        }
    }

    /// Begin a new read. No-op unless current state is
    /// ``ReaderUiState/idle``.
    ///
    /// The stream factory is invoked synchronously; the resulting
    /// ``EmvReader`` AsyncStream is consumed on the main actor and
    /// each ``ReaderState`` is mapped into the UI state.
    public func startScan() {
        guard case .idle = uiState else { return }
        uiState = .scanning
        let stream = readerFactory()
        activeTask?.cancel()
        activeTask = Task { [weak self] in
            await self?.consume(stream)
        }
    }

    /// Reset to the initial state after a Done/Failed terminal.
    public func reset() {
        activeTask?.cancel()
        activeTask = nil
        uiState = Self.initialState(availability: availability)
    }

    private func consume(_ stream: AsyncStream<ReaderState>) async {
        for await state in stream {
            uiState = map(state)
        }
    }

    private func map(_ state: ReaderState) -> ReaderUiState {
        switch state {
        case .tagDetected: return .tagDetected
        case .selectingPpse: return .selectingPpse
        case .selectingAid(let aid): return .selectingApplication(aidHex: String(describing: aid).uppercased())
        case .readingRecords: return .readingRecords
        case .done(let card): return .done(card: EmvCardSummary.from(card))
        case .failed(let error): return mapFailure(error)
        }
    }

    private func mapFailure(_ error: ReaderError) -> ReaderUiState {
        return .failed(
            error: error,
            friendlyMessage: ErrorMessages.friendly(for: error),
            diagnostic: ErrorMessages.diagnostic(for: error)
        )
    }

    private func isResettable(_ state: ReaderUiState) -> Bool {
        switch state {
        case .nfcUnavailable, .idle: return true
        default: return false
        }
    }

    private static func initialState(availability: NfcAvailability) -> ReaderUiState {
        switch availability.current() {
        case .available: return .idle
        case .unavailable: return .nfcUnavailable
        }
    }
}

/// Friendly + diagnostic strings for ``ReaderError`` variants.
///
/// Mirrors the Android composeApp `ErrorPanel.friendlyMessage`. Kept
/// `internal` so tests can drive the same mapping without going through
/// the view-model.
internal enum ErrorMessages {

    static func friendly(for error: ReaderError) -> String {
        switch error {
        case .ioFailure(let reason): return ioMessage(reason)
        case .ppseUnsupported: return "This card does not support contactless EMV."
        case .ppseRejected: return "The card's application directory could not be parsed."
        case .gpoRejected: return "The GET PROCESSING OPTIONS response could not be parsed."
        case .selectAidFciRejected: return "The application FCI could not be parsed."
        case .pdolRejected: return "The card's PDOL could not be parsed."
        case .apduStatusError(let sw1, let sw2):
            return String(format: "Card returned status %02X %02X. Try a different card.", sw1, sw2)
        case .parseFailed: return "The card data could not be parsed into an EMV card."
        }
    }

    static func diagnostic(for error: ReaderError) -> String {
        switch error {
        case .ioFailure(let reason): return "ioFailure(\(reason))"
        case .ppseUnsupported: return "ppseUnsupported"
        case .ppseRejected(let cause): return "ppseRejected(\(String(describing: cause)))"
        case .gpoRejected(let cause): return "gpoRejected(\(String(describing: cause)))"
        case .selectAidFciRejected(let cause): return "selectAidFciRejected(\(String(describing: cause)))"
        case .pdolRejected(let cause): return "pdolRejected(\(String(describing: cause)))"
        case .apduStatusError(let sw1, let sw2):
            return String(format: "apduStatusError(SW1=%02X, SW2=%02X)", sw1, sw2)
        case .parseFailed(let cause): return "parseFailed(\(String(describing: cause)))"
        }
    }

    private static func ioMessage(_ reason: IoReason) -> String {
        switch reason {
        case .tagLost: return "Card moved away from the device. Try again."
        case .timeout: return "The card took too long to respond. Try again."
        case .generic: return "Communication with the card failed. Try again."
        }
    }
}
